creme/neighbors/base_neighbors.py

```diff
@@ -0,0 +1,187 @@
+    that when size limit is reached, old samples are dumped to give
+    place to new samples.
+
+    The internal buffer does not keep order of the stored samples,
+    when the size limit is reached, the older samples are overwritten
+    with new ones (circular buffer).
+
+    """
+
+    def __init__(self, window_size=1000):
+        self.window_size = window_size
+        self._n_features = -1
+        self._n_targets = -1
+        self._size = 0
+        self._next_insert = 0
+        self._oldest = 0
+        self._imask = None
+        self._X = None
+        self._y = None
+        self._is_initialized = False
+
+    def _configure(self):
+        # Binary instance mask to filter data in the buffer
+        self._imask = np.zeros(self.window_size, dtype=bool)
+        self._X = np.zeros((self.window_size, self._n_features))
+        self._y = [None for _ in range(self.window_size)]
+        self._is_initialized = True
+
+    def reset(self):
+        """Reset the sliding window. """
+        self._n_features = -1
+        self._n_targets = -1
+        self._size = 0
+        self._next_insert = 0
+        self._oldest = 0
+        self._imask = None
+        self._X = None
+        self._y = None
+        self._is_initialized = False
+
+        return self
+
+    def add_one(self, x, y):
+        """Add a (single) sample to the sample window.
+
+        x : numpy.ndarray of shape (1, n_features)
+            1D-array of feature for a single sample.
+
+        y : numpy.ndarray of shape (1, n_targets)
+            1D-array of targets for a single sample.
+
+        Raises
+        ------
+        ValueError: If at any moment, a sample with a different number of
+        attributes than that of the n_attributes parameter is passed, a
+        ValueError is raised.
+
+        TypeError: If the buffer type is altered by the user, or is not
+        correctly initialized, a TypeError may be raised.
+
+        """
+        if not self._is_initialized:
+            self._n_features = get_dimensions(x)[1]
+            self._n_targets = get_dimensions(y)[1]
+            self._configure()
+
+        if self._n_features != get_dimensions(x)[1]:
+            raise ValueError("Inconsistent number of features in X: {}, previously observed {}.".
+                             format(get_dimensions(x)[1], self._n_features))
+
+        self._X[self._next_insert, :] = x
+        self._y[self._next_insert] = y
+
+        slot_replaced = self._imask[self._next_insert]
+
+        # Update the instance storing logic
+        self._imask[self._next_insert] = True  # Mark slot as filled
+        self._next_insert = self._next_insert + 1 if self._next_insert < self.window_size - 1 \
+            else 0
+
+        if slot_replaced:  # The oldest sample was replaced (complete cycle in the buffer)
+            self._oldest = self._next_insert
+        else:  # Actual buffer increased
+            self._size += 1
+
+    def remove_one(self, rem_oldest=True):
```

Comment:
I would call this `popleft` to follow the `collections.deque` terminology.

Reply:
Changed in [1a91c21](https://github.com/creme-ml/creme/pull/323/commits/1a91c2109076c5b00b947242115ed4f346681c42).